risc0/zkvm/src/prove/exec.rs

```diff
@@ -362,245 +257,126 @@ impl<'a, H: HostHandler> CircuitStepHandler<BabyBearElem> for MachineContext<'a,
-                    for j in 0..DIGEST_WORDS {
-                        faults.include(info, addr1 + (j * WORD_SIZE) as u32, IncludeDir::Read);
-                        faults.include(info, addr2 + (j * WORD_SIZE) as u32, IncludeDir::Read);
-                    }
-                }
-            } else if minor == ecall::SOFTWARE {
-                let out_addr = self.memory.load_register(REG_A0);
-                let out_words = self.memory.load_register(REG_A1);
-                let out_bytes = out_words * WORD_SIZE as u32;
-
-                for addr in (out_addr..(out_addr + out_bytes)).step_by(WORD_SIZE) {
-                    faults.include(info, addr, IncludeDir::Both);
-                }
-            } else if minor == ecall::HALT {
-                let mode = self.memory.load_register(REG_A0);
-                if mode == halt::PAUSE {
-                    faults.force_flush = true;
-                }
-            }
+    fn page_info(&mut self, _pc: Elem) -> (Elem, Elem, Elem) {
+        if let Some(page_idx) = self.faults.reads.pop_last() {
+            return (Elem::ONE, page_idx.into(), Elem::ZERO);
         }
 
-        faults
-    }
-
-    fn page_info(&mut self, pc: BabyBearElem) -> (BabyBearElem, BabyBearElem, BabyBearElem) {
-        let pc: u32 = pc.into();
-        let inst = self.memory.load_u32(pc);
-        let opcode = OpCode::decode(inst);
-        let info = self.get_page_faults(pc, inst, &opcode);
-
         if self.is_flushing {
-            if let Some(page_idx) = self.dirty_pages.pop_first() {
-                // let info = &self.memory.ram.borrow().info;
-                // debug!(
-                //     "dirty> page_idx: 0x{page_idx:08X}, page_addr: {:08X}",
-                //     info.get_page_addr(page_idx)
-                // );
-                return (BabyBearElem::ZERO, page_idx.into(), BabyBearElem::ZERO);
-            }
-        } else {
-            for page_idx in info.reads.iter().rev() {
-                if !self.finished_page_reads.contains(page_idx) {
-                    self.finished_page_reads.insert(*page_idx);
-                    // debug!("read> page_idx: 0x{page_idx:08X}");
-                    return (BabyBearElem::ONE, (*page_idx).into(), BabyBearElem::ZERO);
-                }
+            if let Some(page_idx) = self.faults.writes.pop_first() {
+                log::debug!("page_write: 0x{page_idx:08x}");
+                return (Elem::ZERO, page_idx.into(), Elem::ZERO);
             }
         }
 
-        (BabyBearElem::ZERO, BabyBearElem::ZERO, BabyBearElem::ONE)
+        (Elem::ZERO, Elem::ZERO, Elem::ONE)
     }
 
-    fn trace(&mut self, cycle: usize, pc: BabyBearElem) -> Result<()> {
-        if self.trace_enabled {
-            let cycle = cycle as u32;
-            let pc = pc.into();
-            self.handler
-                .on_trace(TraceEvent::InstructionStart { cycle, pc })?
-        }
+    fn trace(&mut self, _cycle: usize, _pc: Elem) -> Result<()> {
```

Comment:
I think this function can also be removed, similar to https://github.com/risc0/risc0/pull/514/files#r1170563305

Reply:
Eventually, yes, but that will require changes to the circuit, which I didn't want to do just yet. I think there's a few minor modifications to the circuit that I plan to make soon.